gsoc2020/20200303-health-check-wjayesh.md

```diff
@@ -0,0 +1,254 @@
+ 
+While CoreDNS exposes a health check itself in the form of Kubernetes’ `livenessProbe`: 
+
+
+- The health check is not UDP (DNS) based. There have been cases where the health port is accessible (for TCP) but CoreDNS itself isn't (UDP). This protocol difference means that CoreDNS is unhealthy from a cluster standpoint, but the control plane can't see this.  
+- The existing health check is also launched locally (the `kubelet` uses the `livenessProbe`) and the situation could be different for pods accessing it remotely. 
+ 
+## Motivation and Scope
+
+
+This project idea aims to get around limitations on Kubernetes’ health check and build an application that: 
+
+
+- Checks CoreDNS health externally through UDP (DNS), from a remote Golang application. 
+- Restart CoreDNS pods by interacting with Kubernetes API through the Golang application, if the response from the cluster and pod IPs is unsatisfactory.  
+
+Thus, making the state of CoreDNS available externally and reliably is important to ensure important services run as they are expected to. A shell script could’ve been used to automate the steps but shell is not good in error handling and may not be reliable depending on which shell (bash/csh/zsh) to use. A Golang program will be much more robust.  
+ 
+### Deliverables  
+
+
+- A standalone Golang application to check CoreDNS health: 
+  - The binary delivered through this project will be a **flexible** application in that it could choose its approach to the problem depending on the environment it is deployed in and the access that has been provided to it.   
+  - The health check program itself doesn’t necessarily reside in the Kubernetes cluster. 
+  - This binary only concerns itself with the health of CoreDNS deployed on Kubernetes. It doesn’t cover other scenarios. 
+
+
+- Proper documentation for the product explaining steps and the concepts used. 
+ 
+ 
+ 
+
+ 
+## Design Proposal 
+
+
+It is necessary to consider all possible cases because the way any team deploys the binary is their **choice** and we should not be forcing a modus operandi to use the product.  
+ 
+Minimal assumptions about the users’ Kubernetes deployment are made and the primary **focus** is to build a program that could easily be adopted by more teams (which may have different preferences of deployment). 
+ 
+In any case, the binary will accomplish the following fundamental tasks: 
+
+
+- Use `kubectl` to extract information about the cluster and its services and pods.  
+ 
+  This will help determine the procedure to follow and also facilitate the subsequent actions to take. This step can be completed only when the application can connect and communicate with the API server. The Go [client](https://github.com/kubernetes/client-go) for Kubernetes has packages that can discover and authenticate with the API server.  
```

Comment:
> This will help determine the procedure to follow and also facilitate the subsequent actions to take.

@WJayesh , this sentence may not be necessary.

> This step can be completed only when the application can connect and communicate with the API server.

@WJayesh ... how about "If the application is unable to reach the cluster, then ..."

> The Go [client](https://github.com/kubernetes/client-go) for Kubernetes has packages that can discover and authenticate with the API server.  

@WJayesh , this belongs to the "Implementation" section.

```
First, the application 
```